token-lending/program/tests/repay.rs

```diff
@@ -140,4 +145,43 @@ async fn test_success() {
         recent_blockhash,
     );
     assert!(banks_client.process_transaction(transaction).await.is_ok());
+
+    TestReturn {
+        banks_client,
+        obligation,
+        usdc_reserve,
+    }
+}
+
+#[tokio::test]
+async fn test_repay_obligation() {
```

Comment:
Why did you decide to split the obligation and repay reserve checks off from the success test?

Reply:
With unit tests, you want to have as few assertions as possible in each test, so that you aren't hunting down line numbers for each test failure and the assertion that's failed is clear and legible. I was trying to split these up a bit since they could be compartmentalized a bit more to be "object-level" checks, checking obligation in one and repay reserve in the other. I find that unit tests, when they get very verbose and "omnipotent", become very hard to maintain and fragile.

Reply:
I agree on your approach to unit tests but these are integration tests which are intended to test end-to-end. Unit tests can be found in `src/state/repay.rs` and `src/state/obligation.rs` and are better suited for what I think you are aiming for

Reply:
It still may be easier to make the integration test itself a larger setup function that runs to completion before we do any sort of assertion - that's what makes it integration. Then break the assertion set into many sub tests, or sub functions that is grouped by some organizing heuristic. Then you just either run those in series or re-run the integration setup multiple times, once for each group as a separate test. Technically takes longer to run but one set can pass while the other can fail and you then can have a more clear test suite (multiple failures and successes vs a single failure or success). These may "look" like unit tests in that they are testing a few assertions instead of many but they're still integration tests in that they're calling the same setup function that runs everything together first before asserting anything. 

All this being said, I totally plan to remove this file from this PR, as you've done a way better job in your PR. I'm focusing on Liquidate now :)

Reply:
> It still may be easier to make the integration test itself a larger setup function that runs to completion before we do any sort of assertion - that's what makes it integration. Then break the assertion set into many sub tests, or sub functions that is grouped by some organizing heuristic. 

This sounds good to me!